File: mobile/modules/tutor-sg-llm-runtime/src/TutorSgLlmRuntimeModule.swift
import ExpoModulesCore
import Foundation

public class TutorSgLlmRuntimeModule: Module {
  private let inferenceQueue = DispatchQueue(label: "com.tutorsg.llm", qos: .userInitiated)
  private var modelLoaded = false; private var loadedModelId: String?; private var loadedTier: String?; private var currentState = "uninitialized"
  public func definition() -> ModuleDefinition {
    Name("TutorSgLlmRuntime")
    Events("onToken", "onError", "onStateChange")
    AsyncFunction("loadModel") { (path: String, tier: String) -> [String: Any] in
      var result: [String: Any] = ["success": false, "modelId": "", "state": "error", "message": ""]
      self.inferenceQueue.sync {
        self.emitState("loading")
        guard FileManager.default.fileExists(atPath: path) else { self.sendEvent("onError", ["code": "MODEL_NOT_FOUND", "message": "Model file not found at \(path)", "recoverable": true]); self.emitState("error"); result = ["success": false, "modelId": "", "state": "error", "message": "Model file not found"]; return }
        // ExecuTorch: let module = try Module(filePath: path)
        self.loadedModelId = (path as NSString).lastPathComponent; self.loadedTier = tier; self.modelLoaded = true; self.currentState = "ready"; self.emitState("ready")
        result = ["success": true, "modelId": self.loadedModelId ?? "", "state": "ready", "message": "OK"]
      }
      return result
    }
    AsyncFunction("generate") { (prompt: String, maxTokens: Int, temperature: Double, topP: Double) -> [String: Any] in
      var result: [String: Any] = ["text": "", "tokenCount": 0, "latencyMs": 0, "tokensPerSecond": 0]
      guard self.currentState == "ready" else { self.sendEvent("onError", ["code": "MODEL_NOT_LOADED", "message": "Call loadModel first. Current state: \(self.currentState)", "recoverable": true]); return result }
      self.inferenceQueue.sync {
        self.emitState("generating"); let startMs = CFAbsoluteTimeGetCurrent() * 1000
        // ExecuTorch: try module.forward(input, streamingHandler: handler)
        let elapsedMs = (CFAbsoluteTimeGetCurrent() * 1000) - startMs; self.currentState = "ready"; self.emitState("ready")
        result = ["text": "", "tokenCount": 0, "latencyMs": Int(elapsedMs), "tokensPerSecond": 0]
      }
      return result
    }
    AsyncFunction("unloadModel") { self.inferenceQueue.sync { self.loadedModelId = nil; self.loadedTier = nil; self.modelLoaded = false; self.currentState = "uninitialized"; self.emitState("uninitialized") } }
    Function("getState") { currentState }
    Function("getLoadedModelId") { loadedModelId ?? "" }
    Function("isReady") { currentState == "ready" }
  }
  private func emitState(_ state: String) { currentState = state; sendEvent("onStateChange", state) }
}

File: mobile/modules/tutor-sg-llm-runtime/src/TutorSgLlmRuntimeModule.kt
package expo.modules.tutorsgllmruntime
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.util.concurrent.Executors
import kotlin.math.min
import kotlin.math.roundToInt

class TutorSgLlmRuntimeModule : Module() {
  private val exec = Executors.newSingleThreadExecutor()
  private var nativeInference: Any? = null
  private var currentState = "uninitialized"
  private var loadedModelId: String? = null
  private var loadedTier: String? = null
  override fun definition() = ModuleDefinition {
    Name("TutorSgLlmRuntime")
    Events("onToken", "onError", "onStateChange")
    AsyncFunction("loadModel") { path: String, tier: String ->
      exec.execute {
        try {
          emitState("loading"); val modelFile = File(path)
          if (!modelFile.exists()) { sendEvent("onError", mapOf("code" to "MODEL_NOT_FOUND", "message" to "Model file not found at $path", "recoverable" to true)); emitState("error"); return@execute }
          if (modelFile.length() == 0L) { sendEvent("onError", mapOf("code" to "MODEL_EMPTY", "message" to "Model file is empty", "recoverable" to true)); emitState("error"); return@execute }
          // LiteRT-LM: com.google.ai.edge.litert.lm.LiteRtLm.create(options)
          loadedModelId = modelFile.name; loadedTier = tier; currentState = "ready"; emitState("ready")
          mapOf("success" to true, "modelId" to (loadedModelId ?: ""), "state" to "ready", "message" to "OK")
        } catch (e: Exception) {
          android.util.Log.e("TutorSgLlmRuntime", "loadModel failed", e)
          sendEvent("onError", mapOf("code" to "LOAD_FAILED", "message" to (e.message ?: "Unknown error"), "recoverable" to true))
          emitState("error"); mapOf("success" to false, "modelId" to (loadedModelId ?: ""), "state" to "error", "message" to (e.message ?: "Unknown error"))
        }
      }
    }
    AsyncFunction("generate") { prompt: String, maxTokens: Int, temperature: Double, topP: Double ->
      exec.execute {
        if (currentState != "ready") { sendEvent("onError", mapOf("code" to "MODEL_NOT_LOADED", "message" to "Call loadModel first. Current state: $currentState", "recoverable" to true)); return@execute }
        try {
          emitState("generating"); val startMs = System.currentTimeMillis()
          // LiteRT-LM: GenerateOptions.builder().setStreamingCallback { partial, done -> ... }.build()
          emitState("ready"); mapOf("text" to "", "tokenCount" to 0, "latencyMs" to (System.currentTimeMillis() - startMs), "tokensPerSecond" to 0)
        } catch (e: Exception) {
          android.util.Log.e("TutorSgLlmRuntime", "generate failed", e)
          sendEvent("onError", mapOf("code" to "GEN_FAILED", "message" to (e.message ?: "Generation failed"), "recoverable" to false))
          emitState("error"); mapOf("text" to "", "tokenCount" to 0, "latencyMs" to 0, "tokensPerSecond" to 0)
        }
      }
    }
    AsyncFunction("unloadModel") { exec.execute { try { nativeInference = null; loadedModelId = null; loadedTier = null; currentState = "uninitialized"; emitState("uninitialized") } catch (e: Exception) { android.util.Log.e("TutorSgLlmRuntime", "unload failed", e) } } }
    Function("getState") { currentState }
    Function("getLoadedModelId") { loadedModelId ?: "" }
    Function("isReady") { currentState == "ready" }
  }
  private fun emitState(state: String) { currentState = state; sendEvent("onStateChange", state) }
}
